Make tests covered by codecov

WilliamJamieson opened the pull request:
<!-- describe the changes comprising this PR here -->
This PR makes the tests covered by codecov so that we can detect when test code is not being executed.

Fixes #210

**Checklist**
- [x] Added entry in `CHANGES.rst` under the corresponding subsection
- [x] updated relevant tests
- [ ] updated relevant documentation
- [ ] ~Passed romancal regression testing on Jenkins / PLWishMaster. Link: https://plwishmaster.stsci.edu:8081/job/RT/job/romancal/XXX/~ Only touches testing configurations.

braingram commented on tests/conftest.py:
```diff
@@ -2,40 +2,9 @@
     return MANIFEST
-
-
-@pytest.fixture(name="set_up_list_of_l2_files")
```
Are the two functions here unused? Looking back at the history they were added in #143 but I'm not seeing where they were used.

WilliamJamieson replied:
They were totally ununsed.

WilliamJamieson replied:
The back history is that this was used by, but never added by the `ModelContainer` which was removed by PR #204. Clearly #204 missed removing this. The #143 PR added it so that the `ModelContainer` tests would actually be run correctly (`pytest` was raising a clear warning about this).

braingram commented on tests/test_maker_utils.py:
```diff
@@ -144,9 +142,6 @@ def mutate_value(value):
         if isinstance(value, Time):
             return value + 1 * u.day
 
-        if isinstance(value, list):
```
The docstring mentions that 'Pure lists need to be preserved'

https://github.com/spacetelescope/roman_datamodels/blob/e9d93fdef8c80f26036bd09a0404d9ba89a157fd/tests/test_maker_utils.py#L136-L148

Is this no longer necessary? I'm not familiar with this code so would you describe why it's no longer needed and if so, update the docstring?

WilliamJamieson replied:
It was never needed.

braingram commented on tests/test_maker_utils.py:
```diff
@@ -47,8 +47,6 @@ def test_no_extra_fields(node_class, manifest):
     schema_keys = set()
     subschemas = [schema]
-    if "allOf" in schema:
-        subschemas.extend(schema["allOf"])
```
I could use help understanding what this does. Would you explain why this isn't needed? I'm thinking it would cover the case where a RAD schema extends an existing schema by doing something like:
```yaml
allOf:
  - $ref: other_schema-1.0.0.yaml
  - extra_stuff..
```
kind of like what is done in [file_date-1.0.0.yaml](https://github.com/spacetelescope/rad/blob/1c7dd7fee74ba7a8b57c6250ed4399e35199a7a8/src/rad/resources/schemas/tagged_scalars/file_date-1.0.0.yaml#L8-L9). I'm not seeing any schema for the nodes tested here that use that pattern. Is there something in the design of the members of NODE_CLASSES that prevents this kind of schema extension (and makes this part of the test unnecessary in addition to being unused)?

WilliamJamieson replied:
Look at the parameterization: https://github.com/spacetelescope/roman_datamodels/blob/e9d93fdef8c80f26036bd09a0404d9ba89a157fd/tests/test_maker_utils.py#L38

The `FileDate` object generated by the `file_date-1.0.0.yaml` is NOT a `TaggedObjectNode` it is a `TaggedScalarNode`.

This test in not applicable to `TaggedScalarNode` objects because by construction they cannot have any `keys`. The same applies to `TaggedScalarNode` objects as well.

WilliamJamieson replied:
In general, the `TaggedObjectNode` objects should not need to have a top-level `allOf` combiner.

braingram replied:
Thanks for the extra details.

I provided the link to file date to provide an example of the pattern. Thanks for the confirmation that as I mentioned the pattern is not currently used in rad schemas that turn into TaggedObjectNodes.

If I add a schema that uses the pattern to rad and a maker utility to roman_datamodels all tests pass without issue if the changes are based off of rad main and roman_datamodels main. If however I base the changes off of the source branch for this PR (with the lines discussed here removed) the roman_datamodels (but not rad) tests fail with:
```
AssertionError: Dummy instance has extra keys: ...
```
due to the `test_no_extra_fields` test failing to step into the allOf at the base of the added schema.

The pattern of extending one schema with an allOf and ref is used extensively in JWST schemas and there appears to be nothing in RAD that prevents this pattern and nothing in roman_datamodels that prevents turning an extended schema into functional TaggedObjectNode.

If this pattern should not be used in roman wouldn't it be preferable to either:
- add a restriction in the rad schema (if possible)
- add a test in rad (several appear to exist to further check schemas)

instead of removing these lines so that a schema that extends another schema fails the `test_no_extra_fields` test?

If so, those changes seem outside the scope of this PR.

WilliamJamieson replied:
> The pattern of extending one schema with an allOf and ref is used extensively in JWST schemas and there appears to be nothing in RAD that prevents this pattern and nothing in roman_datamodels that prevents turning an extended schema into functional TaggedObjectNode.
> 
> If this pattern should not be used in roman wouldn't it be preferable to either:
> 
> * add a restriction in the rad schema (if possible)
> * add a test in rad (several appear to exist to further check schemas)
> 
> instead of removing these lines so that a schema that extends another schema fails the `test_no_extra_fields` test?
> 
> If so, those changes seem outside the scope of this PR.

As it stands, virtually all changes to the structure of a schema in RAD will naturally induce failures in `roman_datamodels` because currently there is not a solution to dynamically generating node objects filled with null data from the schemas. Effectively, this means any change such as you describe is expected to induce failures in `roman_datamodels`. Thus it is expected that all changes in RAD are developed alongside changes to `roman_datamodels`.

If there becomes a case where an `allOf` combiner is needed in a `TaggedObjectNode`'s schema then this can be re-introduced. However, there is not currently a use case for this and it actively discourages unnecessary usage of a top level `allOf` combiner. I have already had to clean up several unnecessary uses of the `allOf` combiner in RAD, for some examples see spacetelescope/rad#267. While these unnecessary uses of the `allOf` do not effect how the schema is applied, the do unnecessarily complicate the schemas, which makes it more difficult for new/external users to understand the structure of their datamodels (the structure is dynamically generated from these schemas meaning the only reference users have for the structure is the schemas, which is already a barrier to some users).

I would rather not have to continually notice and fix small issues like spacetelescope/rad#267, but it is involved to add such a test in RAD. This is a relatively simple thing to address here, and as already stated there is an implied expectation of co-developing most changes significant changes between RAD and `roman_datamodels` meaning the home of such checks does not matter.

All this being said I will revoke this change as there are clear objections to it.

braingram replied:
Thanks for the clarification. I don't have any objection to preventing the extension of existing schemas (but I don't also have as much involvement in this project as @ddavis-stsci @PaulHuwe @schlafly etc).

Thanks for helping me understand the impact of removing these lines of code. I don't think this change should be in a PR changing the coverage measurement to include tests.

braingram reviewed (changes requested):
Thanks for the changes and adding the tests to the coverage measurement.

I have a few questions about the changes to the tests and would like to hear if the other maintainers have any comments about turning all warnings into errors in the test suite (that seems like an improvement to me).

zacharyburnett reviewed (approved):
looks like coverage is working and reporting, so this LGTM

WilliamJamieson commented:
> This is not universally a good idea. 

I strongly disagree with this statement in the context of test suites. It is generally considered best practice to turn warnings into errors within test suites. There are several reasons for this including:

1. Warnings are generally signals to downstream developers that there is or will be a potential issue.
2. It is quite easy to ignore even rather serious warnings that really ought to be addressed, see https://github.com/spacetelescope/romancal/actions/runs/5506145209/jobs/10034448151#step:10:170. 
3. Test suites are intended to discover issues and potential issues with code prior to it being released.

That is we should have to explicitly address (and think about) any issues or potential issues our test suite.

Organizations such as `astropy` have learned to do this "the hard way" by having to deal with the fallout of not directly addressing the issues flagged by warnings until they start causing real problems. Resolving this fallout required a significant effort from several developers, which could have instead been addressed piecemeal without sudden significant efforts.

> We can get warnings from the time package when leap second information is not available. 

The EFRA warnings about leap seconds we were seeing were the result of the fact that the times we were choosing where too far in the future for there to exist a leap second table to cover those times. As the reason for this is well understood, it is entirely reasonable to filter these warnings out. This does not prevent us from ignoring or catching warnings, it simply turns unhandled warnings into errors so that we are forced to document/track or address them up front.

>We can also get deprecation warnings while we have a version pinned and waiting for our dependencies to upgrade their packages.

Deprecation warnings are the main reason for why warnings in test suites should be errors. They exist to signal to downstream code (such as our code) that something will break in the future. As such we should be forced to address and document these issues. Since warnings do not cause failures, the CI cannot directly signal this issue, which then requires someone to notice and flag these issues.

PaulHuwe commented:
I also agree with changing warnings to errors for tests, with the proviso that we can specify certain warnings to ignore (e.g. leap seconds).

WilliamJamieson commented:
> I also agree with changing warnings to errors for tests, with the proviso that we can specify certain warnings to ignore (e.g. leap seconds).

Yes we can absolutely do that, for example `astropy` ignores several: https://github.com/astropy/astropy/blob/98b399d52d730d91c2cd29babe1e8545caff219c/setup.cfg#L133-L154. 

As far as the leap second warnings, those were resolved via removing the `factories`. The `maker_utils` pick a time default which does not raise the leap second warning. However, we can add this warning to be ignored if it comes up again in addition to any other warnings we actively decide that we can ignore.

braingram reviewed (approved):
Thanks! LGTM
